### PR TITLE
Created TF module to perform AMI backups using Lamda Functions

### PR DESCRIPTION
**Pull request by const-bon**
## What
* Create AMIs periodically for a backup purpose

## Why
* We will have option to backup and restore EC2 instances state fast

## References
* Based on https://gist.github.com/bkozora/724e01903a9ad481d21e

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    ]
+
+    resources = [
+      "arn:aws:logs:*:*:*",
+    ]
+  }
+
+  statement {
+    actions = [
+      "ec2:*",
```
Comment (osterman) [inline main.tf]: Lets use more precise ACLs

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+resource "aws_iam_role_policy" "ami_backup" {
+  name   = "${module.label.id}"
+  role   = "${aws_iam_role.ami_backup.id}"
+  policy = "${data.aws_iam_policy_document.ami_backup.json}"
+}
+
+resource "aws_lambda_function" "ami_backup" {
+  filename         = "${path.module}/lambda_ami_backups.zip"
+  function_name    = "${module.label_backup.id}"
+  description      = "Automatically backs up instances tagged with backup: true"
```
Comment (osterman) [inline main.tf]: typo

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      region    = "${var.region}"
+      ami_owner = "${var.ami_owner}"
+    }
+  }
+}
+
+resource "aws_lambda_function" "ami_cleanup" {
+  filename         = "${path.module}/lambda_ami_cleanups.zip"
+  function_name    = "${module.label_cleanup.id}"
+  description      = "Cleans up old AMI backups"
```
Comment (osterman) [inline main.tf]: Cleanup old AMI backups

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    variables = {
+      region    = "${var.region}"
+      ami_owner = "${var.ami_owner}"
+    }
+  }
+}
+
+resource "aws_cloudwatch_event_rule" "ami_backup" {
+  name                = "${module.label_backup.id}"
+  description         = "Schedule for ami snapshot backups"
```
Comment (osterman) [inline main.tf]: AMI

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  schedule_expression = "${var.backup_schedule}"
+}
+
+resource "aws_cloudwatch_event_rule" "ami_cleanup" {
+  name                = "${module.label_cleanup.id}"
+  description         = "Schedule for ami snapshot cleanup"
```
Comment (osterman) [inline main.tf]: AMI

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  schedule_expression = "${var.cleanup_schedule}"
+}
+
+resource "aws_cloudwatch_event_target" "ami_backup" {
+  rule      = "${aws_cloudwatch_event_rule.ami_backup.name}"
+  target_id = "schedule_ami_backups"
```
Comment (osterman) [inline main.tf]: I think this needs to be namespaced

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  arn       = "${aws_lambda_function.ami_backup.arn}"
+}
+
+resource "aws_cloudwatch_event_target" "ami_cleanup" {
+  rule      = "${aws_cloudwatch_event_rule.ami_cleanup.name}"
+  target_id = "schedule_ami_cleanups"
```
Comment (osterman) [inline main.tf]: Does this need to be namespaced?

File: main.tf (hunk 1 above)
Comment (const-bon) [reply main.tf]: Added
```
      "ec2:DescribeInstances",
      "ec2:CreateImage",
      "ec2:CreateTags"
```

File: main.tf (hunk 2 above)
Comment (const-bon) [reply main.tf]: Replaced with `Automatically backup instances tagged with 'backup: true'`

File: main.tf (hunk 3 above)
Comment (const-bon) [reply main.tf]: Replaced with `Cleanup old AMI backups`

File: main.tf (hunk 4 above)
Comment (const-bon) [reply main.tf]: Replaced with `Schedule for AMI snapshot backups`

File: main.tf (hunk 5 above)
Comment (const-bon) [reply main.tf]: Replaced with `Schedule for AMI snapshot cleanup`

File: main.tf (hunk 6 above)
Comment (const-bon) [reply main.tf]: Replaced with `target_id = "${module.label_backup.id}"`

File: main.tf (hunk 7 above)
Comment (const-bon) [reply main.tf]: Replaced with `target_id = "${module.label_cleanup.id}"`

**Comment by const-bon** [comment]
Applied all required fixes.
@osterman please check.

File: main.tf [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+resource "aws_iam_role_policy" "ami_backup" {
+  name   = "${module.label.id}"
+  role   = "${aws_iam_role.ami_backup.id}"
+  policy = "${data.aws_iam_policy_document.ami_backup.json}"
+}
+
+resource "aws_lambda_function" "ami_backup" {
+  filename         = "${path.module}/ami_backup.zip"
+  function_name    = "${module.label_backup.id}"
+  description      = "Automatically backup instances tagged with 'backup: true'"
```
Comment (osterman) [inline main.tf]: `Snapshot: true`

**Review by osterman — APPROVED** [review]

**Pull request merged**